Casmpet 5341 1.0 addremovereplace ncn (#1208)
Cray-HPE/docs-csm: operations/node_management/Add_Remove_Replace_NCNs.md

# Alpha Framework to Add, Remove, Replace or Move NCNs

> Add, Remove, Replace or Move worker, storage or master node (NCN). Use this procedure in the event that:

Reviewer: ```suggestion
Add, remove, replace, or move NCNs. This applies to worker, storage, or master nodes. Use this procedure in the event that:
```

> - Worker, storage or master nodes are being replaced and the MAC address is changing.

Reviewer: ```suggestion
- Worker, storage, or master nodes are being replaced and the MAC address is changing.
```

- Worker or storage nodes are being added.

> - Worker, storage or master nodes are being moved to a different cabinet.

Reviewer: ```suggestion
- Worker, storage, or master nodes are being moved to a different cabinet.
```

The following workflows are available:

- [Prerequisites](#prerequisites)
- [Add Worker, Storage or Master NCNs](#add-worker-storage-master)
   - [Add NCN Prerequisites](#add-ncn-prerequisites)
   - [Add NCN Procedure](#add-ncn-procedure)
- [Remove Worker, Storage or Master NCNs](#remove-worker-storage-master)
   - [Remove NCN Prerequisites](#remove-ncn-prerequisites)
   - [Remove NCN Procedure](#remove-ncn-procedure)
- [Replace or Move Worker, Storage or Master NCNs](#replace-worker-storage-master)

<a name="prerequisites"></a>
### Prerequisites

The system is fully installed and has transitioned off of the LiveCD.

All activities required for site maintenance are complete.

The latest docs-csm RPM has been installed on the master nodes.

> 1. Run ncn_add_remove_replace_ncn_pre-req.py to adjust the network.

Reviewer: ```suggestion
1. Run `ncn_add_remove_replace_ncn_pre-req.py` to adjust the network.
```

   ```bash
   ncn-m# cd /usr/share/doc/csm/scripts/operations/node_management/Add_Remove_Replace_NCNs/
   ncn-m# ./ncn_add_remove_replace_ncn_pre-req.py 
   ```

>    1. Script will ask 3 questions:

Reviewer: If we want this as a set of sub-steps, we should be more explicit about what the admin is doing. For example:

---
1. Respond to any applicable questions returned after running the script.
    
    The script will ask three questions:
    
    1. How many NCNs would you like to remove?  Do not include NCNs to be add or moved.

    2. How many NCNs would you like to move? Do not include NCNs to be add or remove.

    3. How many NCNs would you like to add? Do not include NCNs to be removed or moved.

>       1. How many NCNs would you like to remove?  Do not include NCNs to be add or moved.

Reviewer: I'm not sure how this looks once you run the script, but it could be helpful to provide some more context on answering these questions. 

Here are some things I was curious about:

Do you input the specific xnames? Or do you just specify the number of NCNs?
Is there anything specific that needs to be done if you are not removing any NCNs?

      2. How many NCNs would you like to move? Do not include NCNs to be add or remove.

      3. How many NCNs would you like to add? Do not include NCNs to be removed or moved.

   2. When adding new NCNs, there will be network configuration changes that will impact changing IPs on computes. __**That will require DVS restart to update the IPs in the DVS node_map.**__

>    3. ncn_add_remove_replace_ncn_pre-req.py will make the network adjustments and will list the xnames that will need to be rebooted after DVS is restarted. See example below:

Reviewer: ```suggestion
   3. `ncn_add_remove_replace_ncn_pre-req.py` will make the network adjustments and will list the xnames that will need to be rebooted after DVS is restarted. See example below:
```

ncn# NODE=ncn-x00n
```

If the component name (xname) is known, set it now. Otherwise it will be determined in a later step.

```bash
ncn# XNAME=<xname>
ncn# echo $XNAME
```

**IMPORTANT:** Ensure the node being added to the system has been properly configured. If the node being added to the system has not been perviously in the system several settings need to be verified. 
*  Ensure that the NCN device to be added has been racked and cabled per the SHCD.

> *  Ensure the NCN BMC is configure with the expected root user credentials.

Reviewer: ```suggestion
*  Ensure the NCN BMC is configured with the expected root user credentials.
```

   
   The NCN BMC credentials needs to match the current global air-cooled BMC default credentials. This can be viewed with the following command:
   ```bash
   ncn-m# VAULT_PASSWD=$(kubectl -n vault get secrets cray-vault-unseal-keys -o json | jq -r '.data["vault-root"]' |  base64 -d)
   ncn-m# kubectl -n vault exec -it cray-vault-0 -c vault -- env \
      VAULT_TOKEN=$VAULT_PASSWD VAULT_ADDR=http://127.0.0.1:8200 \
      vault kv get secret/reds-creds/defaults
   ```

   Example output:
   ```bash
   ==== Data ====
   Key     Value
   ---     -----
   Cray    map[password:foobar username:root] 
   ```

> *  If adding an NCN that was not previously in the system, follow the [Access and Update the Settings for Replacement NCNs](Access_and_Update_the_Settings_for_Replacement_NCNs.md).

Reviewer: ```suggestion
*  If adding an NCN that was not previously in the system, follow the [Access and Update the Settings for Replacement NCNs](Access_and_Update_the_Settings_for_Replacement_NCNs.md) procedure.
```

> *  Ensure the NCN BMC is configured to use DHCP. (This does not apply to the BMC for ncn-m001 since it is statically configured for the site.)

Reviewer: ncn-m001 -> `ncn-m001`

*  Ensure the NCN is configured to boot over the PCIe NICs instead of the Onboard 1 Gig NICs using the [Switch PXE Boot from Onboard NIC to PCIe](../../instal/../install/switch_pxe_boot_from_onboard_nic_to_pcie.md) procedure.

*  If adding an HPE NCN, ensure IPMI is enabled.

   1. Check to see if IPMI is enabled:
      ```bash
      ncn# export IPMI_PASSWORD=changeme
      ncn# curl -k -u root:$IPMI_PASSWORD https://NCN_NODE-mgmt/redfish/v1/Managers/1/NetworkProtocol | jq .IPMI
      ```

      Expected output:
      ```json
      {
         "Port": 623,
         "ProtocolEnabled": true
      }
      ```

>    2. If disabled is disabled, then enable IPMI:

Reviewer: Should this say:

If IPMI is disabled...

      ```bash
      ncn# curl -k -u root:$IPMI_PASSWORD -X PATCH \
         -H 'Content-Type: application/json' \
         -d '{"IPMI": {"Port": 623, "ProtocolEnabled": true}}' \
         https://NCN_NODE-mgmt/redfish/v1/Managers/1/NetworkProtocol | jq
      ```

      Expected output:
      ```
      {
         "error": {
            "code": "iLO.0.10.ExtendedInfo",
            "message": "See @Message.ExtendedInfo for more information.",
            "@Message.ExtendedInfo": [
               {
               "MessageId": "iLO.2.14.ResetRequired"
               }
            ]
         }
      }
      ```

>    3. If disabled IPMI was disabled, then restart the BMC:

Reviewer: ```suggestion
   3. If IPMI was disabled, then restart the BMC:
```

ncn# XNAME=$(ssh $NODE cat /etc/cray/xname)
ncn# echo $XNAME
```
<a name="remove-ncn-procedure"></a>
### Remove NCN Procedure

The following is a high-level overview of the remove NCN workflow:

1. [Remove NCN from Role, Wipe the Disks and Power Down](Add_Remove_Replace_NCNs/Remove_NCN_from_Role.md)

2. [Remove NCN data](Add_Remove_Replace_NCNs/Remove_NCN_Data.md) from SLS, BSS and HSM

3. [Remove Switch Config](Add_Remove_Replace_NCNs/Remove_Switch_Config.md)

4. [Redeploy Services](Add_Remove_Replace_NCNs/Redeploy_Services.md)

5. [Validate Health](Add_Remove_Replace_NCNs/Validate_Health.md)

**IMPORTANT:** Update the SHCD to remove the device. This is only needed if no NCN device will be added back to same location with the same cabling.

<a name="replace-worker-storage-master"></a>
## Replace or Move Worker, Storage or Master NCNs

> Replacing an NCN is defined as removing an NCN of a given type and adding a different NCN of the same type but with different MAC addresses back into the same cabinet slot.

Reviewer: ```suggestion
Replacing an NCN is defined as removing an NCN of a given type and adding a different NCN of the same type, but with different MAC addresses back into the same cabinet slot.
```
